[SIL] Fix linkage of external property wrapper backing initializers

hamishknight opened the pull request:
Refactor the linkage computation for a SILDeclRef, and fix the linkage for external property wrapper backing initializers such that it depends on the access level of the function it's attached to.

rdar://100240367
Resolves #61233

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hborla reviewed (approved):
Thank you for fixing this!

hamishknight committed:
[SIL] Refactor linkage computation

- Factor out the adding of 'external' for non-definitions
- Factor out computation of the linkage limit

Intended to be NFC.

hamishknight committed:
[SIL] Fix linkage of external property wrapper backing initializers

This should be based on the access level of the
function they're attached to.

rdar://100240367

hamishknight commented:
@swift-ci please test
